File: packages/clients/js/src/referent.ts
import * as tg from "./index.ts";

export type Referent<T> = {
	node: T;
	options?: tg.Referent.Options;
};

export namespace Referent {
	export type Options = {
		artifact?: tg.Artifact.Id | null;
		id?: tg.Object.Id | null;
		location?: tg.Location | null;
		name?: string | null;
		path?: string | null;
		tag?: tg.Tag | null;
		tokens?: tg.Authorization.Tokens | null;
	};

	export let withNodeAndToken = <T>(
		node: T,
		token: tg.Authorization.Token | null,
	): tg.Referent<T> => {
		return withNodeAndTokens(node, tg.Authorization.Tokens.withLocal(token));
	};

	export let withNodeAndTokens = <T>(
		node: T,
		tokens: tg.Authorization.Tokens,
	): tg.Referent<T> => {
		let referent: tg.Referent<T> = { node };
		if (Object.keys(tokens).length > 0) {
			referent.options = { tokens };
		}
		return referent;
	};

	export let toData = <T, U>(
		value: tg.Referent<T>,
		f: (node: T) => U,
	): tg.Referent.Data<U> => {
		let node = f(value.node);
		let options: tg.Referent.Data.Options = {};
		if (
			value.options?.artifact !== undefined &&
			value.options.artifact !== null
		) {
			options.artifact = value.options.artifact;
		}
		if (value.options?.id !== undefined && value.options.id !== null) {
			options.id = value.options.id;
		}
		if (
			value.options?.location !== undefined &&
			value.options.location !== null
		) {
			options.location = tg.Location.toDataString(value.options.location);
		}
		if (value.options?.name !== undefined && value.options.name !== null) {
			options.name = value.options.name;
		}
		if (value.options?.path !== undefined && value.options.path !== null) {
			options.path = value.options.path;
		}
		if (value.options?.tag !== undefined && value.options.tag !== null) {
			options.tag = value.options.tag;
		}
		if (value.options?.tokens !== undefined && value.options.tokens !== null) {
			options.tokens = value.options.tokens;
		}
		return {
			node,
			options,
		};
	};

	export let fromData = <T, U>(
		data: tg.Referent.Data<T>,
		f: (node: T) => U,
	): tg.Referent<U> => {
		tg.assert(typeof data === "object");
		let node = f(data.node);
		let options: tg.Referent.Options = {};
		if (
			data.options?.artifact !== undefined &&
			data.options.artifact !== null
		) {
			options.artifact = data.options.artifact;
		}
		if (data.options?.id !== undefined && data.options.id !== null) {
			options.id = data.options.id;
		}
		if (
			data.options?.location !== undefined &&
			data.options.location !== null
		) {
			options.location = tg.Location.fromDataString(data.options.location);
		}
		if (data.options?.name !== undefined && data.options.name !== null) {
			options.name = data.options.name;
		}
		if (data.options?.path !== undefined && data.options.path !== null) {
			options.path = data.options.path;
		}
		if (data.options?.tag !== undefined && data.options.tag !== null) {
			options.tag = data.options.tag;
		}
		if (data.options?.tokens !== undefined && data.options.tokens !== null) {
			options.tokens = data.options.tokens;
		}
		return {
			node,
			options,
		};
	};

	export let toDataString = <T, U extends string>(
		value: tg.Referent<T>,
		f: (node: T) => U,
	): string => {
		let node = f(value.node);
		let string = node.toString();
		let params = [];
		if (
			value.options?.artifact !== undefined &&
			value.options.artifact !== null
		) {
			params.push(`artifact=${encodeURIComponent(value.options.artifact)}`);
		}
		if (value.options?.id !== undefined && value.options.id !== null) {
			params.push(`id=${encodeURIComponent(value.options.id)}`);
		}
		if (
			value.options?.location !== undefined &&
			value.options.location !== null
		) {
			let location = tg.Location.toDataString(value.options.location);
			params.push(`location=${encodeURIComponent(location)}`);
		}
		if (value.options?.name !== undefined && value.options.name !== null) {
			params.push(`name=${encodeURIComponent(value.options.name)}`);
		}
		if (value.options?.path !== undefined && value.options.path !== null) {
			params.push(`path=${encodeURIComponent(value.options.path)}`);
		}
		if (value.options?.tag !== undefined && value.options.tag !== null) {
			params.push(`tag=${encodeURIComponent(value.options.tag)}`);
		}
		for (let [location, token] of Object.entries(value.options?.tokens ?? {})) {
			params.push(
				`tokens[${encodeURIComponent(location)}]=${encodeURIComponent(token)}`,
			);
		}
		if (params.length > 0) {
			string += "?";
			string += params.join("&");
		}
		return string;
	};

	export let fromDataString = <T extends string, U>(
		data: string,
		f: (node: T) => U,
	): tg.Referent<U> => {
		let [nodeString, params] = data.split("?");
		let node = f(nodeString! as T);
		let options: tg.Referent.Options = {};
		if (params !== undefined) {
			for (let param of params.split("&")) {
				let [key, value] = param.split("=");
				if (value === undefined) {
					throw new Error("missing value");
				}
				switch (key) {
					case "artifact": {
						options.artifact = decodeURIComponent(value);
						break;
					}
					case "id": {
						options.id = decodeURIComponent(value);
						break;
					}
					case "location": {
						options.location = tg.Location.fromDataString(
							decodeURIComponent(value),
						);
						break;
					}
					case "name": {
						options.name = decodeURIComponent(value);
						break;
					}
					case "path": {
						options.path = decodeURIComponent(value);
						break;
					}
					case "tag": {
						options.tag = decodeURIComponent(value);
						break;
					}
					default: {
						let match = key?.match(/^tokens\[(.*)\]$/);
						if (match === null || match === undefined) {
							throw new Error("invalid key");
						}
						options.tokens ??= {};
						options.tokens[decodeURIComponent(match[1]!)] =
							decodeURIComponent(value);
					}
				}
			}
		}
		let referent: tg.Referent<U> = {
			node,
			options,
		};
		return referent;
	};

	export let withoutToken = <T>(value: tg.Referent<T>): tg.Referent<T> => {
		let referent: tg.Referent<T> = {
			node: value.node,
		};
		if (value.options !== undefined) {
			referent.options = { ...value.options };
			delete referent.options.tokens;
		}
		return referent;
	};

	export let withoutRuntime = <T>(value: tg.Referent<T>): tg.Referent<T> => {
		let referent = withoutToken(value);
		if (referent.options !== undefined) {
			delete referent.options.location;
		}
		return referent;
	};

	export type Data<T> =
		| string
		| {
				node: T;
				options?: tg.Referent.Data.Options;
		  };

	export namespace Data {
		export type Options = {
			artifact?: tg.Artifact.Id | null;
			id?: tg.Object.Id | null;
			location?: string | null;
			name?: string | null;
			path?: string | null;
			tag?: tg.Tag | null;
			tokens?: tg.Authorization.Tokens | null;
		};
	}
}

File: packages/clients/rust/src/referent.rs
use {
	crate::prelude::*,
	std::path::{Path, PathBuf},
	tangram_uri::Uri,
	tangram_util::serde::is_default,
};

#[derive(
	Clone,
	Debug,
	Eq,
	Hash,
	Ord,
	PartialEq,
	PartialOrd,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
pub struct Referent<T> {
	#[tangram_serialize(id = 0)]
	pub node: T,

	#[serde(default, skip_serializing_if = "is_default")]
	#[tangram_serialize(default, id = 1, skip_serializing_if = "is_default")]
	pub options: Options,
}

#[derive(
	Clone,
	Debug,
	Default,
	Eq,
	Hash,
	Ord,
	PartialEq,
	PartialOrd,
	serde::Deserialize,
	serde::Serialize,
	tangram_serialize::Deserialize,
	tangram_serialize::Serialize,
)]
pub struct Options {
	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 4, skip_serializing_if = "Option::is_none")]
	pub artifact: Option<tg::artifact::Id>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 0, skip_serializing_if = "Option::is_none")]
	pub id: Option<tg::object::Id>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 6, skip_serializing_if = "Option::is_none")]
	pub location: Option<tg::Location>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 3, skip_serializing_if = "Option::is_none")]
	pub name: Option<String>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 1, skip_serializing_if = "Option::is_none")]
	pub path: Option<PathBuf>,

	#[serde(default, skip_serializing_if = "Option::is_none")]
	#[tangram_serialize(default, id = 2, skip_serializing_if = "Option::is_none")]
	pub tag: Option<tg::Specifier>,

	#[serde(default, skip_serializing_if = "tg::authorization::Tokens::is_empty")]
	#[tangram_serialize(
		default,
		id = 5,
		skip_serializing_if = "tg::authorization::Tokens::is_empty"
	)]
	pub tokens: tg::authorization::Tokens,
}

impl<T> Referent<T> {
	pub fn new(node: T, options: Options) -> Self {
		Self { node, options }
	}

	pub fn with_node(node: T) -> Self {
		Self {
			node,
			options: Options::default(),
		}
	}

	#[must_use]
	pub fn with_node_and_token(node: T, token: Option<tg::authorization::Token>) -> Self {
		Self::with_node_and_tokens(node, tg::authorization::Tokens::with_local(token))
	}

	#[must_use]
	pub fn with_node_and_tokens(node: T, tokens: tg::authorization::Tokens) -> Self {
		let options = Options {
			tokens,
			..Default::default()
		};
		Self::new(node, options)
	}

	pub fn node(&self) -> &T {
		&self.node
	}

	pub fn options(&self) -> &Options {
		&self.options
	}

	pub fn artifact(&self) -> Option<&tg::artifact::Id> {
		self.options.artifact.as_ref()
	}

	pub fn id(&self) -> Option<&tg::object::Id> {
		self.options.id.as_ref()
	}

	#[must_use]
	pub fn location(&self) -> Option<&tg::Location> {
		self.options.location.as_ref()
	}

	pub fn name(&self) -> Option<&str> {
		self.options.name.as_deref()
	}

	pub fn path(&self) -> Option<&Path> {
		self.options.path.as_deref()
	}

	pub fn tag(&self) -> Option<&tg::Specifier> {
		self.options.tag.as_ref()
	}

	pub fn token(&self) -> Option<&tg::authorization::Token> {
		self.options.tokens.local()
	}

	pub fn tokens(&self) -> &tg::authorization::Tokens {
		&self.options.tokens
	}

	pub fn replace<U>(self, node: U) -> (tg::Referent<U>, T) {
		(
			tg::Referent {
				node,
				options: self.options,
			},
			self.node,
		)
	}

	pub fn map<U>(self, f: impl FnOnce(T) -> U) -> tg::Referent<U> {
		tg::Referent {
			node: f(self.node),
			options: self.options,
		}
	}

	pub fn try_map<U, E>(self, f: impl FnOnce(T) -> Result<U, E>) -> Result<tg::Referent<U>, E> {
		Ok(tg::Referent {
			node: f(self.node)?,
			options: self.options,
		})
	}

	pub fn inherit<U>(&mut self, parent: &tg::Referent<U>) {
		self.options.inherit(&parent.options);
	}

	#[must_use]
	pub fn without_token(&self) -> Self
	where
		T: Clone,
	{
		let mut referent = self.clone();
		referent.options.tokens.clear();

		referent
	}

	#[must_use]
	pub fn without_location_and_tokens(&self) -> Self
	where
		T: Clone,
	{
		let mut referent = self.clone();
		referent.options.clear_location_and_tokens();

		referent
	}
}

impl<T> Referent<T>
where
	T: std::fmt::Display,
{
	pub fn to_uri(&self) -> Uri {
		let path = self.node.to_string();
		let mut builder = Uri::builder().path(&path);
		if self.options != Options::default() {
			builder = builder
				.query_params(&self.options)
				.map_err(|error| tg::error!(!error, "failed to serialize the query params"))
				.unwrap();
		}
		builder.build().unwrap()
	}
}

impl<T> Referent<T>
where
	T: std::str::FromStr,
{
	pub fn with_uri(uri: &Uri) -> tg::Result<Self> {
		let node = uri
			.path()
			.parse()
			.map_err(|_| tg::error!("failed to parse the node"))?;
		let options = uri
			.query_raw()
			.map(serde_qs::from_str::<Options>)
			.transpose()
			.map_err(|error| tg::error!(!error, "failed to deserialize the query params"))?
			.unwrap_or_default();
		Ok(Self { node, options })
	}
}

impl Options {
	pub fn clear_location_and_tokens(&mut self) {
		self.location = None;
		self.tokens.clear();
	}

	pub fn with_path(path: impl Into<PathBuf>) -> Self {
		Self {
			artifact: None,
			id: None,
			location: None,
			name: None,
			path: Some(path.into()),
			tag: None,
			tokens: tg::authorization::Tokens::default(),
		}
	}

	pub fn inherit(&mut self, parent: &Options) {
		self.tokens.inherit(&parent.tokens);
		if self.location.is_none() {
			self.location.clone_from(&parent.location);
		}
		if self.id.is_none() && self.tag.is_none() {
			self.id = parent.id.clone();
			self.tag = parent.tag.clone();
			match (&self.path, &parent.path) {
				(None, Some(parent_path)) => {
					let path = parent_path.clone();
					self.path = Some(path);
				},
				(Some(self_path), Some(parent_path)) => {
					let path = parent_path.parent().unwrap().join(self_path);
					let path = tangram_util::path::normalize(&path);
					self.path = Some(path);
				},
				_ => (),
			}
		}
	}
}

impl<T> std::fmt::Display for Referent<T>
where
	T: std::fmt::Display,
{
	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
		write!(f, "{}", self.to_uri())
	}
}

impl<T> std::str::FromStr for Referent<T>
where
	T: std::str::FromStr,
{
	type Err = tg::Error;

	fn from_str(value: &str) -> tg::Result<Self, Self::Err> {
		let uri = Uri::parse(value).map_err(|error| tg::error!(!error, "invalid uri"))?;
		let reference = Self::with_uri(&uri)?;
		Ok(reference)
	}
}

#[cfg(test)]
mod tests {
	use {crate::prelude::*, std::collections::BTreeMap};

	#[test]
	fn location_and_tokens_roundtrip() {
		let id: tg::file::Id = "fil_010000000000000000000000000000000000000000000000000000"
			.parse()
			.unwrap();
		let token = tg::authorization::Token {
			body: tg::authorization::token::Body {
				expires_at: i64::MAX,
				permissions: vec![tg::authorization::Permission::Object(
					tg::authorization::permission::object::Permission::Subtree,
				)],
				resource: id.clone().into(),
			},
			metadata: tg::authorization::token::Metadata {
				algorithm: tg::authorization::token::Algorithm::Ed25519,
				key: "default".into(),
			},
			signature: Vec::new(),
		};
		let remote = tg::Location::Remote(tg::location::Remote {
			name: "default".into(),
			region: None,
		});
		let tokens = tg::authorization::Tokens(BTreeMap::from([
			(
				tg::Location::Local(tg::location::Local::default()),
				token.clone(),
			),
			(remote.clone(), token),
		]));
		let options = tg::referent::Options {
			location: Some(remote),
			tokens,
			..tg::referent::Options::default()
		};
		let referent = tg::Referent::new(id, options);
		let string = referent.to_string();
		let parsed: tg::Referent<tg::file::Id> = string.parse().unwrap();

		assert_eq!(referent, parsed);
		let referent = referent.without_location_and_tokens();
		assert!(referent.options.location.is_none());
		assert!(referent.options.tokens.is_empty());
	}
}
